### DISCUSSION TITLE
webpack 5.44.0 Watch mode: Cannot read property 'get' of undefined

### DISCUSSION THREAD

**Discussion by gkn** [webpack/webpack#13776] [General]
**Do you want to request a *feature* or report a *bug*?**

bug

**What is the current behavior?**

This is from a very large codebase for a web application.
Initial builds are running OK. When running webpack watch it will fail compiling with the following message when changing a source file:

[1] ERROR in [initial]
[1] Cannot read property 'get' of undefined
[1] TypeError: Cannot read property 'get' of undefined
[1]     at C:\Source\cdc\spotless\src\Front\ko\node_modules\webpack\lib\asset\AssetModulesPlugin.js:183:29
[1]     at Hook.eval [as call] (eval at create (C:\Source\cdc\spotless\src\Front\ko\node_modules\webpack\node_modules\tapable\lib\HookCodeFactory.js:19:10), <anonymous>:12:16)
[1]     at Compilation.getRenderManifest (C:\Source\cdc\spotless\src\Front\ko\node_modules\webpack\lib\Compilation.js:3771:36)
[1]     at C:\Source\cdc\spotless\src\Front\ko\node_modules\webpack\lib\Compilation.js:3790:22
[1]     at symbolEach (C:\Source\cdc\spotless\src\Front\ko\node_modules\neo-async\async.js:2444:9)
[1]     at Object.each (C:\Source\cdc\spotless\src\Front\ko\node_modules\neo-async\async.js:2849:16)
[1]     at Compilation.createChunkAssets (C:\Source\cdc\spotless\src\Front\ko\node_modules\webpack\lib\Compilation.js:3784:12)
[1]     at C:\Source\cdc\spotless\src\Front\ko\node_modules\webpack\lib\Compilation.js:2548:14
[1]     at C:\Source\cdc\spotless\src\Front\ko\node_modules\webpack\lib\Compilation.js:2680:5
[1]     at C:\Source\cdc\spotless\src\Front\ko\node_modules\neo-async\async.js:2818:7
[1]
[1] webpack 5.44.0 compiled with 1 error in 3314 ms
[1] i ｢wdm｣: Failed to compile.

**If the current behavior is a bug, please provide the steps to reproduce.**

This used to work fine up to and inclusive of webpack version 5.41.1
On version 5.42.0 I would get similar errors in watch mode, but some of these seem to have been fixed in 5.42.1
Now I get the above message.

```javascript
// webpack.js extracts
...
  cache: {
    type: "filesystem",
  },
...
  optimization: {
    moduleIds: "deterministic",
    runtimeChunk: "single",
    splitChunks: {
      cacheGroups: {
        vendor: {
          test: /[\\/]node_modules[\\/]/,
          name: "vendors",
          chunks: "all",
        },
      },
    },
  },
...

```

**What is the expected behavior?**

**If this is a feature request, what is motivation or use case for changing the behavior?**

**Please mention other relevant information such as the browser version, Node.js version, webpack version, and Operating System.**

**Comment by alexander-akait** [comment]
Can you provide more information? Maybe you can provide list of plugins? How you run webpack, do you use webpack-dev-server?

**Reply by gkn** [reply]
Yes, I use webpack-dev-server: `webpack serve`

We also found a work-around, ref comment https://github.com/webpack/webpack/discussions/13776#discussioncomment-997335

**Comment by accwill** [comment]
I have the same problem

**Comment by DanSnow** [comment]
I got into this yesterday. But I tried several methods e.g. disabling plugins, updating config to fix it. Thus, I'm not sure how did I fix it. The last thing I did is **delete the cache under `node_modules/.cache`**. Now I can't reproduce it.

I have a project running on beta version `5.0.0-beta.2` of `vue-cli` and it is using webpack `5.44.0`. I migrated the `url-loader` & `file-loader` to `Asset Module` with `asset` then I got this error. After that, I tried:
1. disable circular-dependency & firendly-errors plugin
2. change `type: 'asset'` to `type: 'asset/resource'`
3. delete cache -> it fixed

Then I re-enable plugins and I haven't got this error again.

Webpack & Plugins List:
- webpack@npm:5.44.0
- @soda/friendly-errors-webpack-plugin@npm:1.8.0
- copy-webpack-plugin@npm:8.1.1
- case-sensitive-paths-webpack-plugin@npm:2.4.0
- circular-dependency-plugin@npm:5.2.2
- fork-ts-checker-webpack-plugin@npm:6.2.12
- html-webpack-plugin@npm:5.3.2
- vue-loader@npm:15.9.0 (`vue-loader` also contain a plugin)

Installed but did not enable Plugins:
- css-minimizer-webpack-plugin@npm:2.0.0
- mini-css-extract-plugin@npm:1.6.0
- terser-webpack-plugin@npm:5.1.3
- webpack-bundle-analyzer@npm:4.4.2

**Comment by gkn** [comment]
@DanSnow Thanks! The change of module.rules.type from `asset` to `asset/resource`did the trick for me too.

```javascript
// webpack.config.js
...
  module: {
    rules: [
...
      {
        test: /\.(png|jpe?g|gif|svg|eot|ttf|woff|woff2)$/i,
        type: "asset/resource",
      },
...
```

The asset sources, in my case, is from the following npm packages:

"bootstrap": "3.4.1"
"font-awesome": "4.7.0"

**Comment by accwill** [comment]
@DanSnow Thanks! The change of module.rules.type from asset to asset/resourcedid the trick for me too.

**Comment by alexander-akait** [comment]
Maybe somebody can provide example of the problem?

**Reply by await-ovo** [reply]
I also encountered this problem in my project with `cache.type: filesystem`,   But I can not find the minimal example that can be reproduced.

**Reply by alexander-akait** [reply]
Maybe you can capture stacktrace (screenshot/video)?

**Reply by gkn** [reply]
@alexander-akait @sokra 

Release 5.46.0 seems to have fixed this issue - at least in my case, i.e. webpack dev server in watch mode.

``` javascript
// webpack.config.js
      {
        test: /\.(png|jpe?g|gif|svg|eot|ttf|woff|woff2)$/i,
        type: "asset", // The "asset/resource" workaround is no longer needed
      },
```

Thanks :)

**Comment by dmaklygin1** [comment]
the issue definitely exists.
it happens, when there are *inline* assets in the node_modules folder.

config:

```
const imagesLoader = {
    test: [/\.bmp$/, /\.gif$/, /\.jpe?g$/, /\.png$/],
    type: 'asset/inline',
    parser: {
      dataUrlCondition: {
        maxSize: imageInlineSizeLimit
      }
    }
  }
```


To reproduce it:
1. run `yarn && yarn start` in your master branch
2. switch to another branch from your master
3. run `yarn && yarn start` — everything should be OK
4. create a new file, like react component and include it 
you will see these errors:

```
ERROR in Cannot read property 'get' of undefined
TypeError: Cannot read property 'get' of undefined
    at XXXX/node_modules/webpack/lib/asset/AssetModulesPlugin.js:193:29

```


it happens, because, first time, the inline assets from the node_modules are determined as a JAVASCRIPT modules, but second time: as ASSETS modules.
this is a module for the inline asset:

```

ASSETS: PRODUCTION:  NormalModule {
  dependencies: [],
  blocks: [],
  type: 'asset/inline',
  context: '[REMOVED]/node_modules/[REMOVED]/Favicon/icons/production',
  layer: null,
  needId: true,
  debugId: 3328,
  resolveOptions: {},
  factoryMeta: undefined,
  useSourceMap: true,
  useSimpleSourceMap: false,
  _warnings: undefined,
  _errors: undefined,
  buildMeta: { exportsType: 'default' },
  buildInfo: {
    cacheable: true,
    parsed: true,
    fileDependencies: undefined,
    contextDependencies: undefined,
    missingDependencies: undefined,
    buildDependencies: undefined,
    valueDependencies: undefined,
    hash: 'ee362aef00610bd621cb032c8acca544',
    assets: undefined,
    assetsInfo: undefined,
    strict: true,
    dataUrl: true,
    snapshot: Snapshot {
      _flags: 769,
      startTime: 1627556221236,
      fileTimestamps: undefined,
      fileHashes: undefined,
      fileTshs: undefined,
      contextTimestamps: undefined,
      contextHashes: undefined,
      contextTshs: undefined,
      missingExistence: undefined,
      managedItemInfo: [Map],
      managedFiles: [Set],
      managedContexts: undefined,
      managedMissing: undefined,
      children: undefined
    }
  },
  presentationalDependencies: undefined,
  request: '[REMOVED]/node_modules/[REMOVED]/Favicon/icons/production/production-16x16.png',
  userRequest: '[REMOVED]/node_modules/[REMOVED]/Favicon/icons/production/production-16x16.png',
  rawRequest: './production-16x16.png',
  binary: true,
  parser: AssetParser { dataUrlCondition: true },
  parserOptions: { dataUrlCondition: { maxSize: 10000 } },
  generator: AssetGenerator {
    dataUrlOptions: { encoding: undefined, mimetype: undefined },
    filename: undefined,
    publicPath: undefined,
    emit: true
  },
  generatorOptions: undefined,
  resource: '[REMOVED]/icons/production/production-16x16.png',
  resourceResolveData: undefined,
  matchResource: undefined,
  loaders: [],
  error: null,
  _source: RawSource {
    _valueIsBuffer: true,
    _value: <Buffer 89 50 4e 47 0d 0a 1a 0a 00 00 00 0d 49 48 44 52 00 00 00 10 00 00 00 10 08 06 00 00 00 1f f3 ff 61 00 00 00 01 73 52 47 42 00 ae ce 1c e9 00 00 00 44 ... 439 more bytes>,
    _valueAsBuffer: <Buffer 89 50 4e 47 0d 0a 1a 0a 00 00 00 0d 49 48 44 52 00 00 00 10 00 00 00 10 08 06 00 00 00 1f f3 ff 61 00 00 00 01 73 52 47 42 00 ae ce 1c e9 00 00 00 44 ... 439 more bytes>
  },
  _sourceSizes: Map(2) { 'javascript' => 42, 'asset' => 1 },
  _sourceTypes: Set(2) { 'javascript', 'asset' },
  _lastSuccessfulBuildMeta: { exportsType: 'default' },
  _forceBuild: false,
  _isEvaluatingSideEffects: false,
  _addedSideEffectsBailout: undefined
}

```

as you can see, there is not information about pathname inside `buildInfo`.

**Reply by dmaklygin1** [reply]
after clearing `node_modules/.cache` the error disappeared, but might appear again in future.

**Comment by billyvg** [comment]
Upgrading to `5.48.0` seems to have fixed it for us.
